File: results/2026-06-30_191904/tasks/12-pr-label-assigner/typescript-bun-sonnet5/generated-code/run-act-tests.ts
#!/usr/bin/env bun
// act integration harness for .github/workflows/pr-label-assigner.yml.
//
// Per the task's "test through the pipeline" requirement, this is where the
// solution is actually verified end-to-end: for each fixture case it builds
// an isolated temp git repo containing the project + that case's changed
// file list, runs `act push --rm` against it, and asserts on the exact
// LABELS= output the workflow computed -- not just that *something* ran.
//
// Usage: bun run run-act-tests.ts
// Exits 0 only if every case passes; non-zero otherwise.
import { spawnSync } from "node:child_process";
import {
  appendFileSync,
  cpSync,
  mkdtempSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";

const PROJECT_ROOT = import.meta.dir;
const RESULT_FILE = join(PROJECT_ROOT, "act-result.txt");

/** One end-to-end pipeline test case. */
interface ActCase {
  /** Short identifier, also used to name the temp dir. */
  name: string;
  /** Fixture file (relative to the project root) to use as the PR's changed files. */
  fixturePath: string;
  /** Exact expected value of the "LABELS=" line the workflow must print. */
  expectedLabels: string;
}

// Expected values were derived by hand from .github/labeler-config.json and
// confirmed against `bun run src/cli.ts --files <fixture>` during
// development (see task notes) -- these fixtures now serve as the pipeline's
// acceptance cases.
const CASES: ActCase[] = [
  {
    name: "api-and-tests",
    fixturePath: "fixtures/cases/api-and-tests.txt",
    // api(10) > tests(5) > code(0): one file matches two rules, priority
    // breaks the multi-rule ordering.
    expectedLabels: "api,tests,code",
  },
  {
    name: "ci-and-dependencies",
    fixturePath: "fixtures/cases/ci-and-dependencies.txt",
    // dependencies(8) > ci(7): two different files, two different rules,
    // priority decides the conflict.
    expectedLabels: "dependencies,ci",
  },
  {
    name: "docs-only",
    fixturePath: "fixtures/cases/docs-only.txt",
    expectedLabels: "documentation",
  },
  {
    name: "no-match",
    fixturePath: "fixtures/cases/no-match.txt",
    expectedLabels: "",
  },
];

// Everything the workflow needs at runtime. No node_modules/bun.lock --
// the script has zero runtime dependencies, so the workflow never runs
// `bun install`.
const COPY_ITEMS = ["src", "fixtures", ".github", "package.json", "tsconfig.json", ".actrc"];

function setupRepo(c: ActCase): string {
  const dir = mkdtempSync(join(tmpdir(), `act-${c.name}-`));
  for (const item of COPY_ITEMS) {
    cpSync(join(PROJECT_ROOT, item), join(dir, item), { recursive: true });
  }
  // Overwrite the default fixture with this case's changed-file list so the
  // workflow's default push-trigger path picks it up.
  cpSync(join(PROJECT_ROOT, c.fixturePath), join(dir, "fixtures/changed-files.txt"));

  const git = (...args: string[]) =>
    spawnSync("git", args, { cwd: dir, encoding: "utf8" });
  git("init", "-q", "-b", "main");
  git("config", "user.email", "act-harness@example.com");
  git("config", "user.name", "act harness");
  git("add", "-A");
  git("commit", "-q", "-m", `fixture: ${c.name}`);
  return dir;
}

/** Strips ANSI color codes act may emit, so string matching is reliable. */
function stripAnsi(s: string): string {
  return s.replace(/\x1b\[[0-9;]*m/g, "");
}

/** Extracts the value of the last "LABELS=..." occurrence in act's output. */
function extractLabels(output: string): string | null {
  const matches = [...stripAnsi(output).matchAll(/LABELS=([^\r\n]*)/g)];
  if (matches.length === 0) return null;
  return matches[matches.length - 1]![1]!.trimEnd();
}

function section(title: string): string {
  const bar = "=".repeat(78);
  return `\n${bar}\n${title}\n${bar}\n`;
}

function runCase(c: ActCase): boolean {
  console.log(`\n>>> act push for case "${c.name}"`);
  const dir = setupRepo(c);
  try {
    const result = spawnSync(
      "act",
      ["push", "--rm", "--pull=false"],
      { cwd: dir, encoding: "utf8", maxBuffer: 64 * 1024 * 1024 },
    );
    const output = `${result.stdout ?? ""}\n${result.stderr ?? ""}`;
    const exitCode = result.status ?? -1;

    appendFileSync(
      RESULT_FILE,
      section(
        `CASE: ${c.name} | fixture=${c.fixturePath} | expected LABELS=${c.expectedLabels}`,
      ) +
        output +
        `\n[harness] act exit code: ${exitCode}\n`,
    );

    const problems: string[] = [];
    if (exitCode !== 0) problems.push(`act exited ${exitCode}, expected 0`);
    if (!/Job succeeded/.test(output)) {
      problems.push('output did not contain "Job succeeded"');
    }
    const labels = extractLabels(output);
    if (labels === null) {
      problems.push("no LABELS= line found in act output");
    } else if (labels !== c.expectedLabels) {
      problems.push(
        `LABELS mismatch: got "${labels}", expected "${c.expectedLabels}"`,
      );
    }

    if (problems.length > 0) {
      console.error(`  FAIL [${c.name}]: ${problems.join("; ")}`);
      appendFileSync(RESULT_FILE, `[harness] RESULT: FAIL — ${problems.join("; ")}\n`);
      return false;
    }
    console.log(`  PASS [${c.name}]: LABELS="${labels}"`);
    appendFileSync(RESULT_FILE, `[harness] RESULT: PASS\n`);
    return true;
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
}

writeFileSync(
  RESULT_FILE,
  "act integration results for .github/workflows/pr-label-assigner.yml\nGenerated by run-act-tests.ts\n",
);

let passed = 0;
for (const c of CASES) {
  if (runCase(c)) passed++;
}

const summary = `\n${"#".repeat(78)}\nSUMMARY: ${passed}/${CASES.length} cases passed\n${"#".repeat(78)}\n`;
appendFileSync(RESULT_FILE, summary);
console.log(summary.trim());
console.log(`Full output written to: ${RESULT_FILE}`);

process.exit(passed === CASES.length ? 0 : 1);

File: results/2026-06-30_191904/tasks/13-dependency-license-checker/default-sonnet5-low/generated-code/run_act_tests.py
#!/usr/bin/env python3
"""
Test harness that runs the dependency-license-checker workflow through
`act` for each fixture-driven test case, appends the raw output to
act-result.txt, and asserts exact expected values in that output.

Each case copies the project into a fresh temp git repo, swaps in the
case's package.json, then runs `act push --rm`.
"""
import json
import os
import shutil
import subprocess
import sys
import tempfile

ROOT = os.path.dirname(os.path.abspath(__file__))
RESULT_FILE = os.path.join(ROOT, "act-result.txt")

PROJECT_FILES = [
    "license_checker.py",
    "fixtures",
    "tests",
    ".github",
]

# Each case overrides fixtures/package.json and states the exact expected
# summary lines that must appear in the act output.
CASES = [
    {
        "name": "all-approved",
        "package_json": {
            "name": "example-app",
            "version": "1.0.0",
            "dependencies": {
                "left-pad": "1.3.0",
                "express": "4.18.2",
            },
            "devDependencies": {
                "jest": "29.0.0",
            },
        },
        "expected_lines": [
            "[APPROVED] left-pad==1.3.0 -> MIT",
            "[APPROVED] express==4.18.2 -> MIT",
            "[APPROVED] jest==29.0.0 -> MIT",
            "Approved: 3",
            "Denied: 0",
            "Unknown: 0",
        ],
    },
    {
        "name": "one-unknown-license",
        "package_json": {
            "name": "example-app",
            "version": "1.0.0",
            "dependencies": {
                "left-pad": "1.3.0",
                "express": "4.18.2",
                "flask": "2.3.2",
            },
            "devDependencies": {
                "jest": "29.0.0",
            },
        },
        "expected_lines": [
            "[UNKNOWN ] flask==2.3.2 -> BSD-3-Clause",
            "Approved: 3",
            "Denied: 0",
            "Unknown: 1",
        ],
    },
]


def setup_case_repo(tmpdir, case):
    for item in PROJECT_FILES:
        src = os.path.join(ROOT, item)
        dst = os.path.join(tmpdir, item)
        if os.path.isdir(src):
            shutil.copytree(src, dst)
        else:
            shutil.copy2(src, dst)

    with open(os.path.join(tmpdir, "fixtures", "package.json"), "w", encoding="utf-8") as f:
        json.dump(case["package_json"], f, indent=2)

    actrc_src = os.path.join(ROOT, ".actrc")
    if os.path.isfile(actrc_src):
        shutil.copy2(actrc_src, os.path.join(tmpdir, ".actrc"))

    subprocess.run(["git", "init", "-q"], cwd=tmpdir, check=True)
    subprocess.run(["git", "config", "user.email", "test@example.com"], cwd=tmpdir, check=True)
    subprocess.run(["git", "config", "user.name", "Test Runner"], cwd=tmpdir, check=True)
    subprocess.run(["git", "add", "-A"], cwd=tmpdir, check=True)
    subprocess.run(["git", "commit", "-q", "-m", f"test case: {case['name']}"], cwd=tmpdir, check=True)


def run_case(case):
    with tempfile.TemporaryDirectory(prefix=f"act-{case['name']}-") as tmpdir:
        setup_case_repo(tmpdir, case)
        proc = subprocess.run(
            ["act", "push", "--rm", "--pull=false"],
            cwd=tmpdir,
            capture_output=True,
            text=True,
            timeout=600,
        )
        return proc.returncode, proc.stdout + proc.stderr


def main():
    all_passed = True
    with open(RESULT_FILE, "w", encoding="utf-8") as out:
        for case in CASES:
            out.write(f"===== TEST CASE: {case['name']} =====\n")
            returncode, output = run_case(case)
            out.write(output)
            out.write(f"\n===== EXIT CODE: {returncode} =====\n\n")

            case_ok = True
            if returncode != 0:
                print(f"[FAIL] {case['name']}: act exited {returncode}")
                case_ok = False

            job_success_count = output.count("Job succeeded")
            if job_success_count < 2:
                print(f"[FAIL] {case['name']}: expected 2 'Job succeeded' lines, found {job_success_count}")
                case_ok = False

            for expected in case["expected_lines"]:
                if expected not in output:
                    print(f"[FAIL] {case['name']}: expected line not found: {expected!r}")
                    case_ok = False

            if case_ok:
                print(f"[PASS] {case['name']}")
            else:
                all_passed = False

    return 0 if all_passed else 1


if __name__ == "__main__":
    sys.exit(main())
